Call NavigationAgent update_navigation in set_target_location

Shou opened the pull request:
We move the NavigationAgent3D and NavigationAgent2D calls to
`update_navigation()` from the getter functions, to
`set_target_location()` such that it is always updated when a new target
location is set.

Test project: [Shou/godot-demo-projects/tree/45962a9be3ae5f2d2a29b285d0d45d206a55340d/2d/navigation_agent](https://github.com/Shou/godot-demo-projects/tree/45962a9be3ae5f2d2a29b285d0d45d206a55340d/2d/navigation_agent)

Toggle the comment for `nav_agent.get_next_location()` in `navigation.gd`. It's derived from the original navigation example hence the stale comments.

Fixes https://github.com/godotengine/godot/issues/59104

Scony commented on scene/2d/navigation_agent_2d.cpp:
```diff
@@ -178,14 +178,14 @@ void NavigationAgent2D::set_target_location(Vector2 p_location) {
 	target_reached = false;
 	navigation_finished = false;
 	update_frame_id = 0;
+	update_navigation();
```
it looks like it did not help

Scony commented on scene/2d/navigation_agent_2d.cpp:
```diff
@@ -178,14 +178,14 @@ void NavigationAgent2D::set_target_location(Vector2 p_location) {
 }
 
 Vector2 NavigationAgent2D::get_next_location() {
-	update_navigation();
```
this introduces regression

Scony commented on scene/2d/navigation_agent_2d.cpp:
```diff
@@ -208,12 +208,10 @@ bool NavigationAgent2D::is_target_reachable() {
 }
 
 bool NavigationAgent2D::is_navigation_finished() {
-	update_navigation();
```
this introduces regression

Scony commented on scene/2d/navigation_agent_2d.cpp:
```diff
@@ -208,12 +208,10 @@ bool NavigationAgent2D::is_target_reachable() {
 }
 
 Vector2 NavigationAgent2D::get_final_location() {
-	update_navigation();
```
this introduces regression

Scony commented on scene/3d/navigation_agent_3d.cpp:
```diff
@@ -185,14 +185,14 @@ void NavigationAgent3D::set_target_location(Vector3 p_location) {
 }
 
 Vector3 NavigationAgent3D::get_next_location() {
-	update_navigation();
```
this introduces regression

Scony commented on scene/3d/navigation_agent_3d.cpp:
```diff
@@ -215,12 +215,10 @@ bool NavigationAgent3D::is_target_reachable() {
 }
 
 bool NavigationAgent3D::is_navigation_finished() {
-	update_navigation();
```
this introduces regression

Scony commented on scene/3d/navigation_agent_3d.cpp:
```diff
@@ -215,12 +215,10 @@ bool NavigationAgent3D::is_target_reachable() {
 }
 
 Vector3 NavigationAgent3D::get_final_location() {
-	update_navigation();
```
this introduces regression

Scony commented on scene/3d/navigation_agent_3d.cpp:
```diff
@@ -185,14 +185,14 @@ void NavigationAgent3D::set_target_location(Vector3 p_location) {
 	target_reached = false;
 	navigation_finished = false;
 	update_frame_id = 0;
+	update_navigation();
```
it looks like it did not help

Scony reviewed (changes requested):
Surprisingly the proposed changes did not help - actually, they made things worse. I write "surprisingly" as I'd make the changes exactly the same as proposed.
Anyway, I've modified the code and the only solution which makes sense is to add `update_navigation()` to the `get_nav_path()` getter. Ofc. At that point that function must be moved to `.cpp` file and `const` must go away.

Shou commented:
Thanks for the review @Scony! It seemed to work locally for me when I compiled master vs my branch, in that on my branch whether or not `nav_agent.get_next_location()` was commented made no difference and it could navigate just fine -- did my example NavigationAgent2D project not cover the changes properly? What was it that didn't work and what was the regression you found, just so I understand what's going on before I make more changes?

Scony commented:
Well, I tested it on my project, on `3.x`. Maybe `3.x` and `master` are impacted in different ways.
Regarding regression - movement stalled after 2nd location change.
And regarding fix - basically, the path was not updated after setting the target location.

Anyway, could you please check if modifying `get_nav_path()` on master is enough to get it working there? If so, that would be better as this would consistently fix the problem both on `master` and `3.x`.

Shou committed:
Call NavigationAgent update_navigation in set_target_location

We add NavigationAgent3D and NavigationAgent2D calls to
`update_navigation` to additonal functions to ensure we don't return
stale navigation data:

- `is_target_reached()`
- `get_nav_path()`
- `get_nav_path_index()`

Shou commented:
Hey @Scony -- sorry for the delay (life's being hectic, but I have some downtime again now) -- I have a branch with the same changes on 3.x and this is the behaviour I'm getting:

https://user-images.githubusercontent.com/819413/160409245-f452adfa-ff5c-43a7-9841-796d062da318.mp4

Godot 3.x built from this: https://github.com/Shou/godot/tree/nav-agent-set-target-update-nav
And with this project: https://github.com/Shou/godot-demo-projects/tree/nav-agent-set-target-update-nav/2d/navigation

These are the same changes as I applied to master as far as I'm aware, is that not working on your end?

In any case I've pushed and tested the changes adding `update_navigation()` to `get_nav_path()`, I'm not 100% sure about these but they look like they're needed too, so I've done `get_nav_path_index()` and `is_target_reached()` because they can both be modified in `update_navigation()` if `reload_path` is set to true or `navigation_finished` is false.

Let me know what you think

alazifk commented:
IMHO the bug is clearly just a missing update of the navigation path. When we request the navigation path, we have to make sure it is up to date, there is no way around that. So if @Scony is encountering issues, I would assume that to be either some sort of setup problem or a different bug, which would be `update_navigation()` itself failing to deliver. But I do not see any room for the call itself to be wrong. (I have done some tests in the master branch without encountering issues.)

I believe `set_target_location()` should not call `update_navigation()`. Instead this should be lazily evaluated when we actually are using the path for something.

akien-mga commented:
CC @smix8

smix8 commented:
@akien-mga 
I commented on the original issue. The agent is not supposed to be used with manual path index movement and the function `get_nav_path()` is not supposed to trigger updates. At least not without a replacement function that does not trigger updates and returns the full path. `update_navigation()` is also not exposed because it is used internally by the agent and should only trigger once each physics frame max.

I don't think this pr is a good idea / proper fix. Primarly because navigation updates are limited to 1 each sync phase / physics tick so spamming update requests with nearly every function call will lead to an important onetime query / update getting lost in the void on top of performance degradation.

Andrea can correct me but the NavigationAgent is designed to set a target location and then use `get_next_location()` calls for path index advancements and this function uses update_navigation() to triggers path updates and repaths. Maybe get_nav_path() is badly named and feeds to much into old user habits with manual advancing paths with path index. If users want only a fresh full path for manual path index movement they don't need the agent, a call to `NavigationServer2D/3D.map_get_path()` will return what they want without any agent delay.

If to many users find the current behaviour confusing maybe the functions should be changed internally so the visible functions do just what most users expect without damaging the agents internal logic.

Asarmir commented:
My NavigationAgent3d has been working since godot 4 alpha 3. When I changed from alpha 8 to 9 it stopped moving. Thinking this was a glitch in the matrix I waited and when 10 came out. I tired again and it still wont moved. I started throwing around print functions and I found set_target_location is returning a null value even so I am sending it a Target.global_transform.origin. I did print get_next_location() and seem to get a correct vector.  So I'm confused as to why it's not setting the target.
